### PR TITLE
netdev: add netdev_register() to keep track of netdev IDs 

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This introduces the new `type` and `index` field to the `netdev` struct.
The `type` field is unique for each driver type and thus driver configuration struct. The `index` is the index of the device in the driver configuration struct.

For this the driver needs to call `netdev_register()` on setup.
Only a few drivers have been converted yet as an example.

This creates a unique and stable identifier for network devices that only depends on the board configuration.

With this we can continue (can be split off to a separate PR) to implement an EUI -> device mapping for boards that provide MAC addresses.

As a third step we can also extend `luid_get_euiXX` to use `luid_custom` for the fall-back case.
This will allow for stable addresses across driver resets and can be easily implemented as a follow-up to this.

### Testing procedure

This is just adding IDs without using them, nothing to test.

### Issues/PRs references
required for #14634

### CODE REVIEW

File: drivers/netdev_common/netdev_common.c [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+#include "net/netdev.h"
+
+void netdev_common_init(struct netdev *dev)
+{
+#if IS_ACTIVE(GNRC_NETIF_SINGLE)
+    (void) dev;
+#else
+    static netdev_id_t id;
+    dev->id = id++;
```
Comment (miri64) [inline drivers/netdev_common/netdev_common.c]: IMHO this should not be dependent on initialization order. An ID that is the same across all compile configurations would be the best IMHO. One idea that comes to mind would be a pair of `uint8_t` the first one representing the device model, the second e.g. representative of the configuration. Both could be some kind of hashes/sum (for the device model e.g. of the name, for the parameters of the param struct).

**Comment by jia200x** [comment]
nice one, we could use this for a "netdev_get_by_id" function.

File: drivers/netdev_common/netdev_common.c (hunk 1 above)
Comment (benpicco) [reply drivers/netdev_common/netdev_common.c]: I've updated (rewritten) the PR to match this.

Now before I update every `setup()` function for every driver, I provided an example for `at86rf2xx`.

File: drivers/include/net/netdev.h [hunk 2]
```diff
@@ -271,6 +271,16 @@ typedef struct netdev netdev_t;
...
 
+/**
+ * @brief   Driver types for netdev.
+ * @{
+ */
+typedef enum {
+    NETDEV_AT86RF215,
+    NETDEV_AT86RF2XX,
+} netdev_type_t;
+/** @] */
```
Comment (jeandudey) [inline drivers/include/net/netdev.h:283]:
```suggestion
/** @} */
```
:wink:

**Comment by benpicco** [comment]
Ok I think I finally found a solution that I'm happy with.

 - uses 2 extra bytes of RAM / netdev
 - no need to update all drivers at once
 - boards can define which EUI providers are connected to which netdev

I can split this PR up, but with this comming all together the `samr21-xpro` and `avr-rss2` will finally use the MAC address printed on their labels in RIOT.

**Comment by miri64** [comment]
Removed from CI queue to favor PR that potentially goes into feature freeze.

File: sys/include/net/eui_provider.h [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+ */
+#define NETDEV_INDEX_ANY    (0xFF)
+
+/**
+ * @brief   Function for providing a EUI-48 to a device
+ */
+typedef int (*netdev_get_eui48_cb_t)(const void *arg, eui48_t *addr);
+
+/**
+ * @brief   Function for providing a EUI-64 to a device
```
Comment (chudov) [inline sys/include/net/eui_provider.h:42]: It would be nice to mention that the function shall return 0 in case of success. Otherwise addr will be rejected.
The same for eui48 function.

Comment (benpicco) [reply sys/include/net/eui_provider.h:49]: Added documentation for the function.

**Comment by chudov** [comment]
@benpicco I have implemented EUI-64 provider for boards/derfmega256. How to give it to you to incldude to this PR?

**Comment by benpicco** [comment]
If you have it in a branch somewhere I can just pull the patch in :)

**Comment by benpicco** [comment]
Ah, already [found it](https://github.com/chudov/RIOT/commit/8e305ba7518a37c98c14b49a5621e65df999f143)

**Comment by chudov** [comment]
> Ah, already [found it](https://github.com/chudov/RIOT/commit/8e305ba7518a37c98c14b49a5621e65df999f143).

Please get the updated version from https://github.com/chudov/RIOT/commit/26cd0596873f699492a5ba7a2b78e337bf7fadcd as it has proper MAC address.

**Comment by miri64** [comment]
Requeueing to get the release out ASAP.

File: drivers/include/at86rf215.h [hunk 4]
```diff
@@ -292,8 +292,9 @@ typedef struct at86rf215 {
  * @param[out] dev_09       sub-GHz device descriptor
  * @param[out] dev_24       2.4 GHz device descriptor
  * @param[in]  params       parameters for device initialization
+ * @param[in]  index        Index of the device in the parameter struct
```
Comment (miri64) [inline drivers/include/at86rf215.h:295]: This documentation needs to be clarified. Currently it could be misunderstood to be the the index in `params`, which could be fatal when used that way. Maybe

```suggestion
 * @param[in]  index        index of @p params in a global parameter struct array.
```

There also needs to be documentation for the case if this does not come from a global struct array (use-case: user does not use auto-init and initializes the device "by hand").

Also, since when do structs have indices? 

(Also also please stick to the capitalization style of the rest of the function's doc)

Comment (benpicco) [reply drivers/include/at86rf215.h]: Ok, fixed the documentation.

File: drivers/include/at86rf215.h [hunk 5]
```diff
@@ -292,8 +292,10 @@ typedef struct at86rf215 {
  * @param[out] dev_09       sub-GHz device descriptor
  * @param[out] dev_24       2.4 GHz device descriptor
  * @param[in]  params       parameters for device initialization
+ * @param[in]  index        index of @p params in a global parameter struct array.
+ *                          if initialized manually, pass a unique identifier instead.
```
Comment (miri64) [inline drivers/include/at86rf215.h:296]: A new sentence should still start with a capital letter though ^^

File: drivers/include/dose.h [hunk 6]
```diff
@@ -175,8 +175,10 @@ typedef struct {
  * @brief   Setup a DOSE based device state
  * @param[out]  dev         Handle of the device to initialize
  * @param[in]   params      Parameters for device initialization
+ * @param[in]   index       index of @p params in a global parameter struct array.
```
Comment (miri64) [inline drivers/include/dose.h:178]: Here the capitalization of the parameter documentation is different.

```suggestion
 * @param[in]   index       Index of @p params in a global parameter struct array.
```

File: drivers/include/net/netdev.h [hunk 7]
```diff
@@ -271,6 +272,19 @@ typedef struct netdev netdev_t;
...
+ * @brief   Driver types for netdev.
+ *          To be used for @see netdev_register and to match EUI providers.
+ * @{
+ */
+typedef enum {
+    NETDEV_ANY = 0,         /**< Will match any device type */
+    NETDEV_AT86RF215,
+    NETDEV_AT86RF2XX,
+    NETDEV_DOSE,
+} netdev_type_t;
```
Comment (miri64) [inline drivers/include/net/netdev.h:285]: Maybe add a comment here that this `enum` may be extended:

```suggestion
    /* add more if needed */
} netdev_type_t;
```

Comment (benpicco) [reply drivers/include/net/netdev.h:286]: added

File: drivers/include/net/netdev.h [hunk 8]
```diff
@@ -290,8 +305,32 @@ struct netdev {
 #ifdef MODULE_L2FILTER
     l2filter_t filter[CONFIG_L2FILTER_LISTSIZE];   /**< link layer address filters */
 #endif
+#if !IS_ACTIVE(GNRC_NETIF_SINGLE)
```
Comment (miri64) [inline drivers/include/net/netdev.h:308]: I don't like this reverse dependency for GNRC in `netdev` here. How about introducing a pseudo-module for these members and the `netdev_register()` function, that is included when `GNRC_NETIF_SINGLE` is not defined?

File: Makefile.dep [hunk 9]
```diff
@@ -184,6 +184,10 @@ ifneq (,$(filter trickle,$(USEMODULE)))
   USEMODULE += xtimer
 endif
 
+ifneq (,$(filter eui_provider,$(USEMODULE)))
+  USEMODULE += luid
+endif
```
Comment (miri64) [inline Makefile.dep:189]: All this `eui_provider` stuff seems out of place in a PR that is about keeping track of `netdev`s. Maybe move to another PR that depends on this PR or is dependent on this PR?

File: drivers/include/net/netdev.h (hunk 8 above)
Comment (benpicco) [reply drivers/include/net/netdev.h:308]: So replace `GNRC_NETIF_SINGLE` with a `netif_single` pseudo-module?
Sounds good, I'll make a separate PR.

Comment (miri64) [reply drivers/include/net/netdev.h:308]: No, I meant a `netdev_register` pseudo-module.

Comment (miri64) [reply drivers/include/net/netdev.h:308]: `netif_single` requires to much rework in the lower layers ;-).

Comment (benpicco) [reply drivers/include/net/netdev.h:308]: Then I don't understand.
How would a `netdev_register` pseudo-module help?
The idea was that we don't need to keep track of netif IDs if there is only a single netif.
But that information is currently only provided by `GNRC_NETIF_SINGLE`

Comment (miri64) [reply drivers/include/net/netdev.h:308]: > How would a `netdev_register` pseudo-module help?

```suggestion
#if IS_USED(MODULE_NETDEV_REGISTER)
```

And have `netdev_register` be

```C
static inline void netdev_register(struct netdev *dev, netdev_type_t type, uint8_t index)
{
#if IS_USED(MODULE_NETDEV_REGISTER)
    dev->type  = type;
    dev->index = index;
#else
    (void) dev;
    (void) type;
    (void) index;
#endif
}
```

This way you can still optimize it out, but you don't have the implicit dependency to GNRC here.

Comment (benpicco) [reply drivers/include/net/netdev.h:308]: So

```Make
ifeq (,$(filter gnrc_netif_single,$(USEMODULE)))
  USEMODULE += netdev_register
endif
```

Comment (miri64) [reply drivers/include/net/netdev.h:308]: Yepp.

File: Makefile.dep (hunk 9 above)
Comment (benpicco) [reply Makefile.dep:194]: Ok, moved that to another PR.
May I squash while I remove those commits from this PR?

Comment (miri64) [reply Makefile.dep:194]: Yepp

**Comment by miri64** [comment]
Please rebase #14632 was merged.

**Comment by benpicco** [comment]
Rebased :wink:

**Comment by benpicco** [comment]
Ping?

**Comment by fjmolinas** [comment]
ping @miri64!

**Comment by benpicco** [comment]
rebased to include #14113

File: Makefile.dep [hunk 10]
```diff
@@ -69,6 +69,11 @@ ifneq (,$(filter gnrc_mac,$(USEMODULE)))
   endif
 endif
 
+# Don't register netdevs if there is only a single one  of them
+ifeq (,$(filter gnrc_netif_single,$(USEMODULE)))
+  USEMODULE += netdev_register
+endif
```
Comment (leandrolanzieri) [inline Makefile.dep:75]: This is pulling `netdev_register` almost always now. Although it is a pseudomodule I think the dependency should be changed. AFIK there is no equivalent to `gnrc_netif_single` for other stacks, right? (@miri64 ?) Should this be added when some `netdev_%` module is used instead?

Comment (leandrolanzieri) [reply Makefile.dep:75]: This is affecting https://github.com/RIOT-OS/RIOT/pull/13565 now, as the unneeded dependency `netdev_register` is pulled.

Comment (benpicco) [reply Makefile.dep:75]: `netdev_register` is a pseudomodule, so it won't do anything on it's own.
If not `netdev_` code is used, this will not pull in any additional code.

Comment (jia200x) [reply Makefile.dep:75]: But it changes the dependency resolution. This can indeed change the binary,

Comment (miri64) [reply Makefile.dep:75]: Isn't `netdev_register` a sub-module to `netdev`? If not, IMHO there is nothing speaking against it being one.

Comment (miri64) [reply Makefile.dep:75]: I guess we just missed a step in this discussion: https://github.com/RIOT-OS/RIOT/pull/13746#discussion_r461458518

Comment (jia200x) [reply Makefile.dep:75]: how to proceed with that?

Comment (benpicco) [reply Makefile.dep:75]: but you just provided the solution yourself https://github.com/RIOT-OS/RIOT/pull/13565#discussion_r497556842

Comment (jia200x) [reply Makefile.dep:75]: I thought it was not intended to have a reverse dependency to `gnrc_netif_single`

Comment (miri64) [reply Makefile.dep:75]: I am fine with a reverse dependency how you implemented it. My original concern was about GNRC code in the network device code.
